citations-impl/impl/src/java/org/sakaiproject/citation/impl/DbCitationService.java

```diff
@@ -323,6 +350,334 @@ else if(value instanceof String)
         }
 
 		/* (non-Javadoc)
+		* @see org.sakaiproject.citation.impl.BaseCitationService.Storage#commitCitationCollectionOrder(org.sakaiproject.citation.api.CitationCollectionOrder)
+		*/
+		protected void commitSection(CitationCollectionOrder citationCollectionOrder)
+		{
+			String orderStatement = "insert into " + m_collectionOrderTableName + " VALUES(?,?,?,?,?)";
+
+			Object[] orderFields = new Object[5];
+			orderFields [0] = citationCollectionOrder.getCollectionId();
+			orderFields [1] = null;
+			orderFields [2] = citationCollectionOrder.getLocation();
+			orderFields [3] = citationCollectionOrder.getSectiontype();
+			orderFields [4] = citationCollectionOrder.getValue();
+
+			m_sqlService.dbWrite(orderStatement, orderFields);
+		}
+
+
+
+		/* (non-Javadoc)
+		* @see org.sakaiproject.citation.impl.BaseCitationService.Storage#commitSubsection(org.sakaiproject.citation.api.CitationCollectionOrder)
+		*/
+		protected void commitSubsection(CitationCollectionOrder citationCollectionOrder)
+		{
+			List<CitationCollectionOrder> citationCollectionOrderList = getNestedCollectionAsList(citationCollectionOrder.getCollectionId());
+
+			// insert subsection into the right position in the tree
+			int subSectionLocation = 1;
+			List<CitationCollectionOrder> citationCollectionOrderToSave = new ArrayList<CitationCollectionOrder>();
+			for (CitationCollectionOrder collectionOrder : citationCollectionOrderList) {
+				if (citationCollectionOrder.getLocation() == subSectionLocation){
+					citationCollectionOrderToSave.add(citationCollectionOrder);
+				}
+				citationCollectionOrderToSave.add(collectionOrder);
+				subSectionLocation++;
+			}
+			if (citationCollectionOrderToSave.size()==citationCollectionOrderList.size()){
+				citationCollectionOrderToSave.add(citationCollectionOrder);
+			}
+
+			// delete all citation orders
+			String statement = "delete from " + m_collectionOrderTableName + " where COLLECTION_ID = ? and SECTION_TYPE is not null";
```

Comment:
Do you have a transaction here to make sure that the delete and inserts are atomic?